crates/bevy_utils/src/intern.rs

```diff
@@ -0,0 +1,310 @@
+        self
+    }
+}
+
+impl<T: ?Sized> Leak<T> for Box<T> {
+    fn leak(self) -> &'static T {
+        Box::leak(self)
+    }
+}
+
+impl<T: Clone + ?Sized> Leak<T> for Cow<'static, T> {
+    fn leak(self) -> &'static T {
+        match self {
+            Cow::Borrowed(value) => value,
+            Cow::Owned(value) => Box::leak(Box::new(value)),
+        }
+    }
+}
+
+/// A thread-safe interner which can be used to create [`Interned<T>`] from `&T`
+///
+/// For details on interning, see [the module level docs](self).
+///
+/// The implementation ensures that two equal values return two equal [`Interned<T>`] values.
+pub struct Interner<T: ?Sized + 'static>(OnceLock<RwLock<HashSet<&'static T>>>);
+
+impl<T: ?Sized> Interner<T> {
+    /// Creates a new empty interner
+    pub const fn new() -> Self {
+        Self(OnceLock::new())
+    }
+}
+
+impl<T: Hash + Eq + ?Sized> Interner<T> {
+    /// Return the [`Interned<T>`] corresponding to `value`.
+    ///
+    /// If it is called the first time for `value`, it will possibly leak the value and return an
+    /// [`Interned<T>`] using the obtained static reference. Subsequent calls for the same `value`
+    /// will return [`Interned<T>`] using the same static reference.
+    pub fn intern<Q: Leak<T>>(&self, value: Q) -> Interned<T> {
+        let lock = self.0.get_or_init(Default::default);
+        {
+            let set = lock.read().unwrap_or_else(PoisonError::into_inner);
+            if let Some(value) = set.get(value.borrow()) {
+                return Interned(*value);
+            }
+        }
+        {
+            let mut set = lock.write().unwrap_or_else(PoisonError::into_inner);
+            if let Some(value) = set.get(value.borrow()) {
+                Interned(*value)
+            } else {
+                let leaked = value.leak();
+                set.insert(leaked);
+                Interned(leaked)
+            }
+        }
+    }
+}
+
+impl<T: ?Sized> Default for Interner<T> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// A type that can provide static references to equal values.
+pub trait StaticRef {
+    /// Returns a static reference to a value equal to `self`, if possible.
+    /// This method is used by [`OptimizedInterner::intern`] to optimize the interning process.
+    ///
+    /// # Invariant
+    ///
+    /// The following invariant must be hold:
+    ///
+    /// `ptr_eq(a.static_ref(), b.static_ref()) == true` if and only if `a == b`
+    ///
+    /// where `ptr_eq` is defined as :
+    /// ```
+    /// fn ptr_eq<T>(x: Option<&'static T>, y: Option<&'static T>) -> bool {
+    ///     match (x, y) {
+    ///         (Some(x), Some(y)) => std::ptr::eq(x, y),
+    ///         _ => false,
+    ///     }
+    /// }
+    /// ```
+    fn static_ref(&self) -> Option<&'static Self>;
+}
+
+/// An optimized [`Interner`] for  types that implement [`StaticRef`].
+pub struct OptimizedInterner<T: ?Sized + 'static>(Interner<T>);
```

Comment:
This is a good optimization, but why is `OptimizedInterner` separate from `Interner`? I can't imagine why you'd want an interner that doesn't perform this optimization. Would it be possible to further simplify this by combining these types, and potentially also combining `StaticRef` with `Leak`?

Reply:
I actually started with `static_ref` being a part of `Leak`, but then I realized that the invariant needs to be hold over all implementations of `Leak` and is therefore a property of the interned type and not of the `Leak` implementation.
Therefore I am against combining `StaticRef` with `Leak`.

Regarding the separation of `OptimizedInterner` and `Interner`, my thought was that it might be useful to create an interner for types that do not implement `StaticRef`, which could be the case for foreign types. But this is more of a hypothetical scenario, so I am open to combining those. If we later want to have an Interner for `str` we could also implement `StaticRef` for `str` instead.

Reply:
I've merged the `OptimizedInterner` into `Interner`.